File: week2/omerSefaYilmaz/longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
    let result = 0;
    let freqMap = {};
    let left = 0;
    let mostFreqChar = 0;
  
    for (let right = 0; right < s.length; right++) {
      let char = s[right];
      if (!freqMap[char]) {
        freqMap[char] = 1;
      } else {
        freqMap[char]++;
      }
  
      mostFreqChar = Math.max(mostFreqChar, freqMap[char]);
      let longestStr = right - left + 1 - mostFreqChar;
  
      if (longestStr > k) {
        freqMap[s[left]]--;
        left++;
      }
  
      result = Math.max(right - left + 1, result);
    }
  
    return result;
  };

File: week2/MarshalLee/LongestRepeatingCharacterReplacement.cs
using System;

public class LongestRepeatingCharacterReplacement
{
    public int CharacterReplacement(string s, int k)
    {
        Dictionary<char, int> charFrequencies = new();
        int result = 0;

        int left = 0;
        int maxFrequency = 0;
        for (int right = 0; right < s.Length; right++)
        {
            charFrequencies[s[right]] = 1 + (charFrequencies.ContainsKey(s[right]) ? charFrequencies[s[right]] : 0);
            maxFrequency = Math.Max(maxFrequency, charFrequencies[s[right]]);

            while ((right - left + 1) - maxFrequency > k)
            {
                charFrequencies[s[left]] -= 1;
                left += 1;
            }

            result = Math.Max(result, right - left + 1);
        }

        return result;
    }
}
